n2survey/read_from_xml.py

```diff
@@ -0,0 +1,268 @@
+import re
+import sys
+from json import dump
+from typing import Dict
+
+import bs4
+from bs4 import BeautifulSoup
+
+
+def _text_refining(s: str) -> str:
+    matched = ["\n", "\t", "\xa0"]
+    res_s = re.sub("|".join(matched), "", s)
+    return res_s
+
+
+def _sectioninfo_text_parse(bsString: str) -> str:
+    text = ""
+    sps = BeautifulSoup(bsString, "html.parser").find_all("p")
+    for sp in sps:
+        if bool(sp.string):
+            text += _text_refining(sp.string)
+    return text
+
+
+def _find_choices(resp: bs4.element.Tag) -> dict:
+    cates = dict()
+    categories = resp.find_all("category")
+    for category in categories:
+        category_lab = category.label.string
+        category_val = category.value.string
+        # A#other case
+        if bool(category.contingentQuestion):
+            category_val = category.contingentQuestion.attrs["varName"]
+            category_lab = _text_refining(
+                category.contingentQuestion.find("text").get_text()
+            )
+        cates[category_lab] = category_val
+    return cates
+
+
+def _integrate_subquestions(question: bs4.element.Tag, section_id: int) -> list:
+    res_subquestions = list()
+    subs = question.find_all("subQuestion")
+    if question.response.category is not None:
+        sub_type = "subquestion choice"
+    else:
+        sub_type = "subquestion text"
+    for sub in subs:
+        res_subquestions.append(
+            {
+                "name": sub["varName"],
+                "question": sub.find("text").get_text(),
+                "description": "",
+                "question type": sub_type,
+                "is_other": False,
+                "is_sub": True,
+                "choices": _find_choices(question.response),
+                "section_id": section_id,
+                "qustion group": question.response.attrs["varName"],
+            }
+        )
+    return res_subquestions
+
+
+def _store_questions(
```

Comment:
It looks like the code can be simplified. But let's fix the main function first

Reply:
ye, new version has been pushed, and if you have time, you can check it. this time I have compared the outcome with file "dummy_data_codeandtext.csv" in data directory. They are nealy same, so you can check it.
b.t.w. sorry this late reply.